### PR TITLE
Feature/robot constructor

### PR DESCRIPTION
**Pull request by jlunenburg**

### CODE REVIEW

File: robot_smach_states/examples/test-detect-faces [hunk 1]
```diff
@@ -7,22 +7,16 @@ import sys
...
 
 # Detect face
+from robot_skills import get_robot
 from robot_smach_states import DetectFace
 
 
 if __name__ == "__main__":
 
     rospy.init_node('test_detect_faces')
     robot_name = sys.argv[1]
```
Comment (MatthijsBurgh) [inline robot_smach_states/examples/test-detect-faces]: Add argument check like other files.

File: robot_smach_states/examples/simple_navigate.py [hunk 2]
```diff
@@ -13,15 +12,8 @@
     rospy.init_node('simple_navigate')
 
     robot_name = sys.argv[1]
```
Comment (MatthijsBurgh) [inline robot_smach_states/examples/simple_navigate.py]: Add argument check like other files.

File: robot_skills/test/test_speech_recognition.py [hunk 3]
```diff
@@ -1,7 +1,7 @@
 #! /usr/bin/env python
 import sys
-
 import rospy
+from robot_skills import get_robot
 
 if len(sys.argv) < 2:
```
Comment (MatthijsBurgh) [inline robot_skills/test/test_speech_recognition.py]: Add argument check like other files to be consistent

File: robot_skills/test/test_audio.py [hunk 4]
```diff
@@ -1,33 +1,24 @@
...
 import sys
+import time
 
+# ROS
 import rospy
 
+# TU/e Robotics
+from robot_skills import get_robot
+
 if len(sys.argv) < 2:
```
Comment (MatthijsBurgh) [inline robot_skills/test/test_audio.py]: Add argument check like other files to be consistent

File: robot_smach_states/src/robot_smach_states/navigation/follow_operator2_0.py [hunk 5]
```diff
@@ -433,19 +433,17 @@ def __init__(self, robot):
 
 
 if __name__ == "__main__":
+
+    from robot_skills import get_robot
+
     if len(sys.argv) > 1:
```
Comment (MatthijsBurgh) [inline robot_smach_states/src/robot_smach_states/navigation/follow_operator2_0.py]: Add argument check like other files to be consistent

File: robot_skills/src/robot_skills/get_robot.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    if name == "amigo":
+        return Amigo()
+    elif name == "hero":
+        return Hero()
+    # elif name == "mockbot":
+    #     return Mockbot()
+    elif name == "sergio":
+        return Sergio()
+    else:
+        raise RuntimeError("Don't know which robot to construct with name {}".format(name))
```
Comment (alberth) [inline robot_skills/src/robot_skills/get_robot.py]: Why not make it a dictionary?
``` python
robots = {'amigo': Amigo,
          'hero': Hero,
#          'mockbot': Mockbot,
          'sergio': Sergio,
         }
robot_class = robots.get(name.lower())
if robot_class:
    return robot_class()
else:
    raise RuntimeError("Don't know which robot to construct with name {}".format(name))
````

File: robot_smach_states/test/designators/empty_spot_designator.py [hunk 7]
```diff
@@ -1,12 +1,15 @@
...
 import robot_smach_states.util.designators as ds
+from robot_skills import get_robot
 
 if __name__ == "__main__":
     rospy.init_node("testdesignator")
-    from robot_skills.amigo import Amigo
-    robot = Amigo()
+
+    robot_name = sys.argv[1] if len(sys.argv) > 1 else "amigo"
+    robot = get_robot(robot_name)
```
Comment (alberth) [inline robot_smach_states/test/designators/empty_spot_designator.py]: I wonder if the default robot selection should also be inside `get_robot`. Current solution is more code, but has the nice property that the default is selectable at every call.

Alternatively, move argument selection into a new function?
```` python
import sys

def get_robot_from_argv(index, default_robot_name='amigo'):
    """
    Construct a robot from the name given in the command-line or from the default robot name.
    :param index: Index in the command-line arguments where a robot name may be available.
    :param default_robot_name: Name of the robot to use if the command line did not contain a name.
    :return: The constructed robot.
    :raise: RunTimeError if no robot could be created.
    """
    if len(sys.argv) > index:
        robot_name = sys.argv[index]
    else:
        robot_name = default_robot_name

    return get_robot(robot_name)
````

File: robot_skills/src/robot_skills/get_robot.py (hunk 6 above)
Comment (jlunenburg) [reply robot_skills/src/robot_skills/get_robot.py]: See a71172c306547eba6d50c057baaef8c0ec10a1d0

File: robot_smach_states/test/designators/empty_spot_designator.py (hunk 7 above)
Comment (jlunenburg) [reply robot_smach_states/test/designators/empty_spot_designator.py]: See a71172c306547eba6d50c057baaef8c0ec10a1d0

File: robot_smach_states/examples/test-detect-faces (hunk 1 above)
Comment (jlunenburg) [reply robot_smach_states/examples/test-detect-faces]: See a71172c306547eba6d50c057baaef8c0ec10a1d0 and 106535870505f785620744139c45820da4f330a4

File: robot_smach_states/examples/simple_navigate.py (hunk 2 above)
Comment (jlunenburg) [reply robot_smach_states/examples/simple_navigate.py]: See a71172c306547eba6d50c057baaef8c0ec10a1d0 and 106535870505f785620744139c45820da4f330a4

File: robot_skills/test/test_speech_recognition.py (hunk 3 above)
Comment (jlunenburg) [reply robot_skills/test/test_speech_recognition.py]: See a71172c306547eba6d50c057baaef8c0ec10a1d0 and 106535870505f785620744139c45820da4f330a4

File: robot_skills/test/test_audio.py (hunk 4 above)
Comment (jlunenburg) [reply robot_skills/test/test_audio.py]: See a71172c306547eba6d50c057baaef8c0ec10a1d0 and 106535870505f785620744139c45820da4f330a4

File: robot_smach_states/src/robot_smach_states/navigation/follow_operator2_0.py (hunk 5 above)
Comment (jlunenburg) [reply robot_smach_states/src/robot_smach_states/navigation/follow_operator2_0.py]: See a71172c306547eba6d50c057baaef8c0ec10a1d0 and 106535870505f785620744139c45820da4f330a4

**Comment by LoyVanBeek** [comment]
Merged into the rgo2019 branch

**Pull request closed without merging**